src/nodes.coffee

```diff
@@ -11,11 +11,14 @@ Error.stackTraceLimit = Infinity
 # Import the helpers we plan to use.
 {compact, flatten, extend, merge, del, starts, ends, some,
 addDataToNode, attachCommentsToNode, locationDataToString,
-throwSyntaxError} = require './helpers'
+throwSyntaxError, replaceUnicodeCodePointEscapes} = require './helpers'
 
 # Functions required by parser.
 exports.extend = extend
 exports.addDataToNode = addDataToNode
+# Convert `String` object back to primitive string now that we've retrieved
+# stowaway extra properties
+exports.toPrimitiveString = (strObject) -> "#{strObject}"
```

Comment:
To continue the `"#{$1}"`/`toPrimitiveString()` thread (which got swallowed up), here's the definition of `toPrimitiveString()`

I added the comment you mentioned here - I agree with how you're thinking about it, it'd be nice to use the `"#{$1}"` syntax directly but it's hard to justify adding a verbose comment every time we use it (there are two places in the grammar as of this branch, but there are others on other branches where this technique has been used). So I think the best way of making it comprehensible and relatively DRY is to make them all use a shared function with a reasonably descriptive name and whose definition provides a more thorough comment

Or the alternative in my mind is to just use the `"#{$1}"` syntax directly and let people scratch their heads a bit?

Reply:
I did a [performance test](https://jsperf.com/interpolated-string-vs-tostring/1) of `` `${strObj}` `` vs `strObj.toString()`. In my browser, `toString()` is 25 _times_ faster. So we should definitely use `toString()` instead of the interpolation/casting syntax, it seems.

That doesn’t settle whether we should keep `toPrimitiveString($1)` and call `toString` inside `toPrimitiveString`, or just call `$1.toString()` directly; the latter has the “where to put the comment” issue again. But `toString` is rather descriptive on its own, so maybe we wouldn’t need a comment? (We should at least put a comment before the first time we call `toString`, even if we don’t repeat the comment every time.) Or we could just use `toString` inside the body of `toPrimitiveString` and be done with it. I don’t have a preference.

Reply:
> `toString()` is 25 times faster

That's interesting

> `toString` is rather descriptive on its own

I agree, updated to use `$1.toString()` directly with a descriptive comment before the first block where it's used in the grammar